lms/djangoapps/grades/rest_api/v1/views.py

```diff
@@ -207,3 +217,307 @@ def _get_course(self, request, course_id):
+        else:
+            raise self.api_error(
+                status_code=status.HTTP_403_FORBIDDEN,
+                developer_message='The course does not exist.',
+                error_code='user_or_course_does_not_exist',
+            )
+
+    def _generate_course_xblock_structure(self, course_enrollments):
+        """ Generate a map of course to course enrollment and problem
+            xblocks for each of the course.
+        """
+        course_enrollment_id_map = {}
+        course_xblock_structure = {}
+        for course_enrollment in course_enrollments:
+            course_enrollment_id_map[str(course_enrollment.course_id)] = course_enrollment
+        for course_key, course_enrollment in course_enrollment_id_map.items():
+            course = get_course(course_enrollment.course_id, depth=4)
+            blocks = self._get_problem_blocks(course)
+            course_xblock_structure[course_key] = {
+                'course_enrollment': course_enrollment,
+                'blocks': blocks,
+                'course': course
+            }
+        return course_xblock_structure
+
+    def _get_problem_blocks(self, course):
+        """ Get a list of problem xblock for the course.
+        """
+        blocks = []
+        for section in course.get_children():
+            for subsection in section.get_children():
+                for vertical in subsection.get_children():
+                    for block in vertical.get_children():
+                        if (block.category == 'problem' and getattr(block, 'has_score', False)):
+                            blocks.append(block)
+        return blocks
+
+    def _get_problem_data(self, course_enrollment, block):
+        """
+        Get problem data from a course enrollment.
+
+        Args:
+        -----
+        course_enrollment: Course Enrollment.
+        block: Xblock to analyze.
+        """
+        problem_data = {
+            'location': str(block.scope_ids.usage_id),
+            'name': block.display_name,
+            'submission_history': [],
+            'data': block.data
+        }
+        csm = StudentModule.objects.filter(
+            module_state_key=block.location,
+            student=course_enrollment.user,
+            course_id=course_enrollment.course_id
+        )
+
+        scores = BaseStudentModuleHistory.get_history(csm)
+        for score in scores:
+            state = score.state
+            if state is not None:
+                state = json.loads(state)
+
+            history_data = {
+                'state': state,
+                'grade': score.grade,
+                'max_grade': score.max_grade
+            }
+            problem_data['submission_history'].append(history_data)
+
+        return problem_data
+
+    def _get_course_data(self, course_key, course_enrollment, course, blocks):
+        """
+        Get course data.
+
+        Params:
+        --------
+
+        course_enrollment (CourseEnrollment): course enrollment
```

Comment:
Why don't we annotate types directly in the method's definition?

Reply:
Not all types can be imported, for example, how do I annotate a course or xblock? So I have annotated what I could and kept the rest in doc. Let me know if that is okay.

Reply:
I am leaving it as is!